// react-native-asset/esm/cli.js
import * as dntShim from "./_dnt.shims.js";
import * as path from "./deps/jsr.io/@std/path/1.1.3/mod.js";
import { parseArgs } from "./deps/jsr.io/@std/cli/1.0.24/parse_args.js";
import { linkAssets } from "./main.js";
export const runCli = async () => {
    const args = await parseArgs(dntShim.Deno.args, {
        alias: {
            a: "assets",
            "ios-a": "ios-assets",
            "android-a": "android-assets",
            p: "path",
            "n-u": "no-unlink",
        },
        collect: ["assets", "ios-assets", "android-assets"],
        string: ["assets", "ios-assets", "android-assets", "path"],
        boolean: ["no-unlink"],
        default: {
            path: dntShim.Deno.cwd(),
            noUnlink: false,
        },
    });
    const reactNativeConfigPath = path.join(args.path, `react-native.config.js`);
    let reactNativeConfigExists = false;
    try {
        const _ = await dntShim.Deno.lstat(reactNativeConfigPath);
        reactNativeConfigExists = true;
    }
    catch (err) {
        if (!(err instanceof dntShim.Deno.errors.NotFound)) {
            throw err;
        }
    }
    const reactNativeConfig = reactNativeConfigExists
        ? (await import(reactNativeConfigPath)).default
        : {};
    const merged = {
        assets: args.assets.length !== 0
            ? args.assets
            : (reactNativeConfig.assets ?? undefined),
        iosAssets: args["ios-assets"].length !== 0
            ? args["ios-assets"]
            : (reactNativeConfig.iosAssets ?? undefined),
        androidAssets: args["android-assets"].length !== 0
            ? args["android-assets"]
            : (reactNativeConfig.androidAssets ?? undefined),
    };
    await linkAssets({
        rootPath: args.path,
        shouldUnlink: !args.noUnlink,
        platforms: {
            ios: {
                enabled: merged.assets !== undefined || merged.iosAssets !== undefined,
                assets: [
                    ...(merged.iosAssets ?? []),
                    ...(merged.assets ?? []),
                ],
            },
            android: {
                enabled: merged.assets !== undefined ||
                    merged.androidAssets !== undefined,
                assets: [
                    ...(merged.androidAssets ?? []),
                    ...(merged.assets ?? []),
                ],
            },
        },
    });
};
